src/GraphQLParser/Visitors/SDLWriter.cs

```diff
@@ -0,0 +1,765 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using GraphQLParser.AST;
+
+namespace GraphQLParser.Visitors
+{
+    /// <summary>
+    /// Prints AST into the provided <see cref="TextWriter"/> as a SDL document.
+    /// </summary>
+    /// <typeparam name="TContext">Type of the context object passed into all VisitXXX methods.</typeparam>
+    public class SDLWriter<TContext> : DefaultNodeVisitor<TContext>
+        where TContext : IWriteContext
+    {
+        /// <inheritdoc/>
+        public override async ValueTask VisitDocument(GraphQLDocument document, TContext context)
+        {
+            if (document.Definitions?.Count > 0)
+            {
+                for (int i = 0; i < document.Definitions.Count; ++i)
+                {
+                    await Visit(document.Definitions[i], context);
+
+                    if (i < document.Definitions.Count - 1)
+                    {
+                        await context.Writer.WriteLineAsync();
+                    }
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public override async ValueTask VisitComment(GraphQLComment comment, TContext context)
+        {
+            int level = GetLevel(context);
+
+            bool needStartNewLine = true;
+            int length = comment.Text.Span.Length;
+            for (int i = 0; i < length; ++i)
+            {
+                if (needStartNewLine)
+                {
+                    await WriteIndent(context, level);
+                    await context.Writer.WriteAsync('#');
+                    needStartNewLine = false;
+                }
+
+                char current = comment.Text.Span[i];
+                switch (current)
+                {
+                    case '\r':
+                        break;
+
+                    case '\n':
+                        await context.Writer.WriteLineAsync();
+                        needStartNewLine = true;
+                        break;
+
+                    default:
+                        await context.Writer.WriteAsync(current);
+                        break;
+                }
+            }
+
+            await context.Writer.WriteLineAsync();
+        }
+
+        /// <inheritdoc/>
+        public override async ValueTask VisitDescription(GraphQLDescription description, TContext context)
```

Comment:
Not all strings can be written in the blockstring format.  For example, a string that starts with a newline.  This method should ideally check to be sure the input string can be formatted as a blockstring without loss of information, and if not, be formatted as a string.

Reply:
> and if not, be formatted as a string.

One-line string? Or are you talking about change description node into comment ?

Reply:
See https://github.com/graphql-dotnet/parser/pull/149#discussion_r748885168

Please show an example here or add a test case yourself. I'm a little confused with all this escaping stuff.